File: internal/discovery/wsdiscovery.go
package discovery

import (
	"context"
	"fmt"
	"log"
	"net"
	"strings"
	"sync"
	"time"

	"github.com/google/uuid"
)

const (
	multicastAddr = "239.255.255.250:3702"
	onvifTypes    = "dn:NetworkVideoTransmitter"
)

type Device struct {
	UUID string
	Name string
	IP   string
	Port int
}

// SharedDiscovery handles WS-Discovery for all devices from a single socket
type SharedDiscovery struct {
	mu      sync.Mutex
	devices []Device
	started bool
}

var shared SharedDiscovery

// Register adds a device and starts the shared listener on first call
func Register(ctx context.Context, dev Device) error {
	shared.mu.Lock()
	shared.devices = append(shared.devices, dev)
	needsStart := !shared.started
	shared.started = true
	shared.mu.Unlock()

	if needsStart {
		return startShared(ctx)
	}

	// Send Hello for newly registered device
	addr, _ := net.ResolveUDPAddr("udp4", multicastAddr)
	conn, err := net.DialUDP("udp4", nil, addr)
	if err == nil {
		conn.Write([]byte(makeHello(dev)))
		conn.Close()
	}

	log.Printf("WS-Discovery registered %s on %s", dev.Name, dev.IP)
	return nil
}

func startShared(ctx context.Context) error {
	addr, err := net.ResolveUDPAddr("udp4", multicastAddr)
	if err != nil {
		return err
	}

	// Listen on all interfaces for multicast
	conn, err := net.ListenMulticastUDP("udp4", nil, addr)
	if err != nil {
		// Fallback: listen on 0.0.0.0
		fallback, err2 := net.ListenUDP("udp4", &net.UDPAddr{Port: 3702})
		if err2 != nil {
			return fmt.Errorf("ws-discovery listen: %w (multicast: %v)", err2, err)
		}
		conn = fallback
	}

	log.Printf("WS-Discovery shared listener started on port 3702")

	// Send Hello for all registered devices
	shared.mu.Lock()
	devs := make([]Device, len(shared.devices))
	copy(devs, shared.devices)
	shared.mu.Unlock()

	for _, dev := range devs {
		hello := makeHello(dev)
		conn.WriteToUDP([]byte(hello), addr)
		log.Printf("WS-Discovery Hello sent for %s", dev.Name)
	}

	go func() {
		defer conn.Close()
		buf := make([]byte, 65536)
		for {
			select {
			case <-ctx.Done():
				// Send Bye for all devices
				shared.mu.Lock()
				allDevs := make([]Device, len(shared.devices))
				copy(allDevs, shared.devices)
				shared.mu.Unlock()
				for _, dev := range allDevs {
					conn.WriteToUDP([]byte(makeBye(dev)), addr)
				}
				return
			default:
			}

			conn.SetReadDeadline(time.Now().Add(2 * time.Second))
			n, remoteAddr, err := conn.ReadFromUDP(buf)
			if err != nil {
				continue
			}

			data := string(buf[:n])
			if strings.Contains(data, "Probe") && !strings.Contains(data, "ProbeMatch") {
				msgID := extractTag(data, "MessageID")
				handleProbe(conn, remoteAddr, msgID)
			}
		}
	}()

	return nil
}

func handleProbe(conn *net.UDPConn, remoteAddr *net.UDPAddr, msgID string) {
	shared.mu.Lock()
	devs := make([]Device, len(shared.devices))
	copy(devs, shared.devices)
	shared.mu.Unlock()

	for _, dev := range devs {
		resp := makeProbeMatch(dev, msgID)
		conn.WriteToUDP([]byte(resp), remoteAddr)
	}
	log.Printf("WS-Discovery: sent %d ProbeMatches to %s", len(devs), remoteAddr)
}

// Start is kept for backward compatibility with Python Docker version
func Start(ctx context.Context, dev Device) error {
	return Register(ctx, dev)
}

func extractTag(xml, tag string) string {
	patterns := []string{
		"<wsa:" + tag + ">",
		"<a:" + tag + ">",
	}
	for _, p := range patterns {
		idx := strings.Index(xml, p)
		if idx >= 0 {
			start := idx + len(p)
			end := strings.Index(xml[start:], "</")
			if end >= 0 {
				return strings.TrimSpace(xml[start : start+end])
			}
		}
	}
	return ""
}

func makeHello(dev Device) string {
	return fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope"
  xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing"
  xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery"
  xmlns:dn="http://www.onvif.org/ver10/network/wsdl">
<s:Header>
  <wsa:MessageID>urn:uuid:%s</wsa:MessageID>
  <wsa:To>urn:schemas-xmlsoap-org:ws:2005:04:discovery</wsa:To>
  <wsa:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/Hello</wsa:Action>
  <d:AppSequence InstanceId="1" MessageNumber="1"/>
</s:Header>
<s:Body><d:Hello>
  <wsa:EndpointReference><wsa:Address>urn:uuid:%s</wsa:Address></wsa:EndpointReference>
  <d:Types>%s</d:Types>
  <d:Scopes>%s</d:Scopes>
  <d:XAddrs>http://%s:%d/onvif/device_service</d:XAddrs>
  <d:MetadataVersion>1</d:MetadataVersion>
</d:Hello></s:Body></s:Envelope>`,
		uuid.New(), dev.UUID, onvifTypes, scopes(dev), dev.IP, dev.Port)
}

func makeProbeMatch(dev Device, relatesTo string) string {
	return fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope"
  xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing"
  xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery"
  xmlns:dn="http://www.onvif.org/ver10/network/wsdl">
<s:Header>
  <wsa:MessageID>urn:uuid:%s</wsa:MessageID>
  <wsa:RelatesTo>%s</wsa:RelatesTo>
  <wsa:To>http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous</wsa:To>
  <wsa:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/ProbeMatches</wsa:Action>
  <d:AppSequence InstanceId="1" MessageNumber="1"/>
</s:Header>
<s:Body><d:ProbeMatches><d:ProbeMatch>
  <wsa:EndpointReference><wsa:Address>urn:uuid:%s</wsa:Address></wsa:EndpointReference>
  <d:Types>%s</d:Types>
  <d:Scopes>%s</d:Scopes>
  <d:XAddrs>http://%s:%d/onvif/device_service</d:XAddrs>
  <d:MetadataVersion>1</d:MetadataVersion>
</d:ProbeMatch></d:ProbeMatches></s:Body></s:Envelope>`,
		uuid.New(), relatesTo, dev.UUID, onvifTypes, scopes(dev), dev.IP, dev.Port)
}

func makeBye(dev Device) string {
	return fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope"
  xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing"
  xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery">
<s:Header>
  <wsa:MessageID>urn:uuid:%s</wsa:MessageID>
  <wsa:To>urn:schemas-xmlsoap-org:ws:2005:04:discovery</wsa:To>
  <wsa:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/Bye</wsa:Action>
</s:Header>
<s:Body><d:Bye>
  <wsa:EndpointReference><wsa:Address>urn:uuid:%s</wsa:Address></wsa:EndpointReference>
</d:Bye></s:Body></s:Envelope>`, uuid.New(), dev.UUID)
}

func scopes(dev Device) string {
	name := strings.ReplaceAll(dev.Name, " ", "%20")
	return fmt.Sprintf("onvif://www.onvif.org/type/video_encoder onvif://www.onvif.org/type/Network_Video_Transmitter onvif://www.onvif.org/hardware/ONVIF-Splitter onvif://www.onvif.org/name/%s", name)
}

File: onvif_splitter/discovery/ws_discovery.py
from __future__ import annotations

import asyncio
import logging
import socket
import struct
import uuid
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..virtual_device import VirtualDevice

log = logging.getLogger(__name__)

MULTICAST_GROUP = "239.255.255.250"
MULTICAST_PORT = 3702

WSD_NS = "http://schemas.xmlsoap.org/ws/2005/04/discovery"
WSA_NS = "http://schemas.xmlsoap.org/ws/2004/08/addressing"
ONVIF_TYPES = "dn:NetworkVideoTransmitter"
ONVIF_DN_NS = "http://www.onvif.org/ver10/network/wsdl"


class WsDiscoveryProtocol(asyncio.DatagramProtocol):
    def __init__(self, device: VirtualDevice):
        self.device = device
        self.transport: asyncio.DatagramTransport | None = None

    def connection_made(self, transport: asyncio.DatagramTransport):
        self.transport = transport

    def datagram_received(self, data: bytes, addr: tuple[str, int]):
        try:
            text = data.decode("utf-8", errors="ignore")
            if "Probe" in text and "ProbeMatch" not in text:
                self._handle_probe(text, addr)
        except Exception:
            log.debug("Error handling WS-Discovery packet", exc_info=True)

    def _handle_probe(self, text: str, addr: tuple[str, int]):
        # Extract MessageID for RelatesTo
        msg_id = ""
        if "<wsa:MessageID>" in text:
            start = text.index("<wsa:MessageID>") + len("<wsa:MessageID>")
            end = text.index("</wsa:MessageID>", start)
            msg_id = text[start:end].strip()
        elif "<a:MessageID>" in text:
            start = text.index("<a:MessageID>") + len("<a:MessageID>")
            end = text.index("</a:MessageID>", start)
            msg_id = text[start:end].strip()

        d = self.device
        response = self._probe_match_xml(msg_id)
        if self.transport:
            self.transport.sendto(response.encode(), addr)
            log.debug("Sent ProbeMatch for %s to %s", d.name, addr)

    def _probe_match_xml(self, relates_to: str) -> str:
        d = self.device
        msg_uuid = f"urn:uuid:{uuid.uuid4()}"
        device_uuid = f"urn:uuid:{d.device_uuid}"
        xaddr = d.service_url("/onvif/device_service")
        name_encoded = d.name.replace(" ", "%20")

        scopes = (
            "onvif://www.onvif.org/type/video_encoder "
            "onvif://www.onvif.org/type/Network_Video_Transmitter "
            "onvif://www.onvif.org/hardware/ONVIF-Splitter "
            f"onvif://www.onvif.org/name/{name_encoded}"
        )

        return f"""<?xml version="1.0" encoding="UTF-8"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope"
  xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing"
  xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery"
  xmlns:dn="{ONVIF_DN_NS}">
  <s:Header>
    <wsa:MessageID>{msg_uuid}</wsa:MessageID>
    <wsa:RelatesTo>{relates_to}</wsa:RelatesTo>
    <wsa:To>http://schemas.xmlsoap.org/ws/2004/08/addressing/role/anonymous</wsa:To>
    <wsa:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/ProbeMatches</wsa:Action>
    <d:AppSequence InstanceId="1" MessageNumber="1"/>
  </s:Header>
  <s:Body>
    <d:ProbeMatches>
      <d:ProbeMatch>
        <wsa:EndpointReference><wsa:Address>{device_uuid}</wsa:Address></wsa:EndpointReference>
        <d:Types>{ONVIF_TYPES}</d:Types>
        <d:Scopes>{scopes}</d:Scopes>
        <d:XAddrs>{xaddr}</d:XAddrs>
        <d:MetadataVersion>1</d:MetadataVersion>
      </d:ProbeMatch>
    </d:ProbeMatches>
  </s:Body>
</s:Envelope>"""

    def hello_xml(self) -> str:
        d = self.device
        msg_uuid = f"urn:uuid:{uuid.uuid4()}"
        device_uuid = f"urn:uuid:{d.device_uuid}"
        xaddr = d.service_url("/onvif/device_service")
        name_encoded = d.name.replace(" ", "%20")

        scopes = (
            "onvif://www.onvif.org/type/video_encoder "
            "onvif://www.onvif.org/type/Network_Video_Transmitter "
            "onvif://www.onvif.org/hardware/ONVIF-Splitter "
            f"onvif://www.onvif.org/name/{name_encoded}"
        )

        return f"""<?xml version="1.0" encoding="UTF-8"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope"
  xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing"
  xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery"
  xmlns:dn="{ONVIF_DN_NS}">
  <s:Header>
    <wsa:MessageID>{msg_uuid}</wsa:MessageID>
    <wsa:To>urn:schemas-xmlsoap-org:ws:2005:04:discovery</wsa:To>
    <wsa:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/Hello</wsa:Action>
    <d:AppSequence InstanceId="1" MessageNumber="1"/>
  </s:Header>
  <s:Body>
    <d:Hello>
      <wsa:EndpointReference><wsa:Address>{device_uuid}</wsa:Address></wsa:EndpointReference>
      <d:Types>{ONVIF_TYPES}</d:Types>
      <d:Scopes>{scopes}</d:Scopes>
      <d:XAddrs>{xaddr}</d:XAddrs>
      <d:MetadataVersion>1</d:MetadataVersion>
    </d:Hello>
  </s:Body>
</s:Envelope>"""

    def bye_xml(self) -> str:
        msg_uuid = f"urn:uuid:{uuid.uuid4()}"
        device_uuid = f"urn:uuid:{self.device.device_uuid}"

        return f"""<?xml version="1.0" encoding="UTF-8"?>
<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope"
  xmlns:wsa="http://schemas.xmlsoap.org/ws/2004/08/addressing"
  xmlns:d="http://schemas.xmlsoap.org/ws/2005/04/discovery">
  <s:Header>
    <wsa:MessageID>{msg_uuid}</wsa:MessageID>
    <wsa:To>urn:schemas-xmlsoap-org:ws:2005:04:discovery</wsa:To>
    <wsa:Action>http://schemas.xmlsoap.org/ws/2005/04/discovery/Bye</wsa:Action>
    <d:AppSequence InstanceId="1" MessageNumber="1"/>
  </s:Header>
  <s:Body>
    <d:Bye>
      <wsa:EndpointReference><wsa:Address>{device_uuid}</wsa:Address></wsa:EndpointReference>
    </d:Bye>
  </s:Body>
</s:Envelope>"""


class WsDiscovery:
    def __init__(self, device: VirtualDevice):
        self.device = device
        self._protocol: WsDiscoveryProtocol | None = None
        self._transport: asyncio.DatagramTransport | None = None

    async def start(self):
        loop = asyncio.get_running_loop()

        # Create UDP socket for multicast
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        try:
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)
        except (AttributeError, OSError):
            pass

        sock.bind(("", MULTICAST_PORT))

        # Join multicast group on the device's IP
        group = socket.inet_aton(MULTICAST_GROUP)
        local = socket.inet_aton(self.device.ip)
        mreq = struct.pack("4s4s", group, local)
        sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)

        # Set outgoing multicast interface
        sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_IF, local)

        self._transport, self._protocol = await loop.create_datagram_endpoint(
            lambda: WsDiscoveryProtocol(self.device),
            sock=sock,
        )

        # Send Hello
        hello = self._protocol.hello_xml()
        self._transport.sendto(hello.encode(), (MULTICAST_GROUP, MULTICAST_PORT))
        log.info("WS-Discovery started for %s on %s", self.device.name, self.device.ip)

    async def stop(self):
        if self._protocol and self._transport:
            bye = self._protocol.bye_xml()
            self._transport.sendto(bye.encode(), (MULTICAST_GROUP, MULTICAST_PORT))
            self._transport.close()
            log.info("WS-Discovery stopped for %s", self.device.name)
